Evaluate and integrate hessians

bergbauer opened the pull request:
This PR implements the evaluation and integration of hessians on faces (currently only for affine faces) in the matrix-free infrastructure.

Part of #10401

Planned follow ups are the implementation for general cell/face type and the integration of hessians for cells.

kronbichler commented:
/rebuild

peterrum commented on include/deal.II/matrix_free/evaluation_kernels.h:
```diff
@@ -2523,6 +2789,50 @@ namespace internal
             }
+        if (hessians == true)
+          {
+            // diagonal part
+            const unsigned int hdim = (dim * (dim + 1)) / 2;
+            for (unsigned int d = 0; d < dim; ++d)
```
Do we need to split this into diagonal and off-diagonal part? Why not simply:
```suggestion
            // diagonal part
            const unsigned int hdim = (dim * (dim + 1)) / 2;
            for (unsigned int d = 0; d < dim + hdim; ++d)
```
and remove the last part?

peterrum replied:
Generally, it might be worth to refactor and modularize this method. The same code is applied on different sub-ranges and different input/output vectors.

bergbauer replied:
It is not possible to do it the way you suggest. Also the diagonal and off diagonal part are scaled by different factors. You need to scale the off diagonal entries by to to take the symmetry of the hessians into account.

peterrum replied:
What scaling do you mean?

kronbichler replied:
I agree with @peterrum here - all you do here is exchange some entries for each derivative component, so there should be no need to treat the first `dim` components differently from the next `hdim-dim` components. (Note that in the code proposed by @peterrum you only need to run up to `hdim`, not `dim + hdim`, as the former already contains diagonals and off-diagonals.

bergbauer replied:
Sorry I was thinking about my splitted for loops in `submit_hessian`. You are right, I changed it.

peterrum replied:
Thanks!

peterrum reviewed (commented):
Looks very good! I only have a few minor comments. Furthermore, could you add a change-log entry!

bergbauer commented on tests/matrix_free/matrix_vector_hessians_faces.cc:
```diff
@@ -0,0 +1,1024 @@
+
+template <int dim>
+void
+test_hessians_only(const unsigned int degree, const dealii::FE_Poly<dim> &fe)
```
I refactored it to only use one function!

bergbauer commented on include/deal.II/matrix_free/mapping_info.templates.h:
```diff
@@ -2477,6 +2491,13 @@ namespace internal
+
+                      if (update_flags_faces & update_jacobian_grads)
+                        {
+                          // TODO: implement jacobian grads for general path
+                          AssertThrow(false, ExcNotImplemented());
```
It is not computing these values since `update_flags_faces & update_jacobian_grads` is always false on faces at the moment. I will implement the general path for faces in a follow up PR.

bergbauer commented on include/deal.II/matrix_free/mapping_info.templates.h:
```diff
@@ -2292,10 +2292,14 @@ namespace internal
         AlignedVector<VectorizedDouble> cell_points(dim * n_mapping_points);
         AlignedVector<VectorizedDouble> face_quads(dim * n_q_points);
         AlignedVector<VectorizedDouble> face_grads(dim * dim * n_q_points);
+        AlignedVector<VectorizedDouble> face_grad_grads(dim * hess_dim *
+                                                        n_q_points);
```
Same applies here. This is just a dummy right now.

bergbauer commented:
Thanks for the review @peterrum ! Ready from my side for another round.

peterrum commented on tests/matrix_free/matrix_vector_hessians_faces.cc:
```diff
@@ -0,0 +1,388 @@
+
+#include "../tests.h"
+
+// tests if applying a matrix vector product with a matrix containing hessians
+// on faces produces the same result with FEFaceEvaluation and FEFaceValues
+// this is checked for different combinations of EvaluationFlags, FE types and
+// polynomial degrees
```
I think normally we tend to write here real sentences.

peterrum commented on doc/news/changes/minor/20210728MaximilianBergbauer:
```diff
@@ -0,0 +1,3 @@
+Added: Evaluation and integration of hessians on faces in the matrix free framework.
```
... as well as here.

peterrum reviewed (commented):
Looks good to me :+1: Two more comments and one question.

kronbichler commented on include/deal.II/matrix_free/evaluation_kernels.h:
```diff
@@ -1542,6 +1570,7 @@ namespace internal
                       nullptr,
                       gradients_quad,
                       nullptr,
+                      nullptr,
```
I think it would make more sense to pass the argument on here, 
```suggestion
                      hessians_quad,
                      nullptr,
```
the you can remove the assert from here and rely on the `FEEvaluationImplCollocation` class.

kronbichler commented on include/deal.II/matrix_free/evaluation_kernels.h:
```diff
@@ -1913,37 +1950,52 @@ namespace internal
                      Number *           gradients_quad,
+                     Number *           hessians_quad,
                      Number *           scratch_data,
                      const bool         evaluate_val,
                      const bool         evaluate_grad,
+                     const bool         evaluate_hessian,
```
This reminds me of the fact that we should convert this to the `EvaluationFlags` argument like for the cell integrators, but we don't need it for this PR.

bergbauer replied:
I totally agree. Having those three bools in all interfaces is just ugly, but this is something for a follow up.

kronbichler commented on include/deal.II/matrix_free/evaluation_kernels.h:
```diff
@@ -1974,19 +2026,21 @@ namespace internal
                                         data.n_q_points_face;
 
-      if (evaluate_grad == false)
+      const auto values_dofs_first = values_dofs;
```
```suggestion
      // keep a copy of the original pointer for the case of the Hessians
      const Number values_dofs_ptr = values_dofs;
```

kronbichler commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -855,6 +857,21 @@ class FEEvaluationBase
                            const unsigned int q_point);
 
+  /**
+   * Write a contribution that is tested by the hessian to the field
```
```suggestion
   * Write a contribution that is tested by the Hessian to the field
```

kronbichler commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -855,6 +857,21 @@ class FEEvaluationBase
 
+  /**
+   * Write a contribution that is tested by the hessian to the field
+   * containing the values on quadrature points with component @p q_point.
```
```suggestion
   * containing the values at quadrature points with component @p q_point.
```
(it might be that we use `on` in other comments making it not your fault, but `at` should be the grammatically correct preposition here.)

kronbichler commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -855,6 +857,21 @@ class FEEvaluationBase
+   * containing the values on quadrature points with component @p q_point.
+   * Access to the same field as through get_hessian(). If applied before the
+   * function FEEvaluation::integrate(EvaluationFlags::hessians) is called,
+   * this specifies what is tested by all basis function hessians on the
```
```suggestion
   * this specifies what is tested by the Hessians of all all basis function on the
```

kronbichler commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -1534,6 +1559,12 @@ class FEEvaluationAccess<dim, 1, Number, is_face, VectorizedArrayType>
   void
   submit_gradient(const gradient_type grad_in, const unsigned int q_point);
 
+  /**
+   * @copydoc FEEvaluationBase<dim,1,Number,is_face>::submit_hessian()
+   */
+  void
+  submit_hessian(const hessian_type hessian_in, const unsigned int q_point);
+
```
Can you move this down to line 1580 below the `get_hessian_diagonal` function, to have similar things grouped together. While there, can you also change the return type of `get_hessian` to `hessian_type`, to make use of the new typedef and create consistent names?

kronbichler commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -1882,6 +1914,12 @@ class FEEvaluationAccess<1, 1, Number, is_face, VectorizedArrayType>
   submit_normal_derivative(const gradient_type grad_in,
                            const unsigned int  q_point);
 
+  /**
+   * @copydoc FEEvaluationBase<1,1,Number,is_face>::submit_hessian()
+   */
+  void
+  submit_hessian(const hessian_type hessian_in, const unsigned int q_point);
+
```
Move below the `get_hessian` functions.

kronbichler commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -5795,7 +5832,7 @@ FEEvaluationBase<dim, n_components_, Number, is_face, VectorizedArrayType>::
 
   // Cartesian cell
-  if (this->cell_type == internal::MatrixFreeFunctions::cartesian)
+  if (!is_face && this->cell_type == internal::MatrixFreeFunctions::cartesian)
```
Good catch!

kronbichler commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -6639,6 +6780,97 @@ FEEvaluationAccess<dim, 1, Number, is_face, VectorizedArrayType>::
 
+template <int dim, typename Number, bool is_face, typename VectorizedArrayType>
+inline DEAL_II_ALWAYS_INLINE void
+FEEvaluationAccess<dim, 1, Number, is_face, VectorizedArrayType>::
+  submit_hessian(const Tensor<2, dim, VectorizedArrayType> hessian_in,
+                 const unsigned int                        q_point)
+{
+  Assert(this->cell != numbers::invalid_unsigned_int, ExcNotInitialized());
+  AssertIndexRange(q_point, this->n_quadrature_points);
```
I think this function is expensive enough on its own that we could simply forward the implementation to the generic function `FEEvaluationBase::submit_hessian`, just as `submit_normal_derivative` does, in order to avoid repeating a lot of code.

kronbichler commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -8399,10 +8646,10 @@ FEEvaluation<dim,
+                          EvaluationFlags::hessians)) == 0,
     ExcMessage(
-      "Only EvaluationFlags::values and EvaluationFlags::gradients are supported."));
+      "Only EvaluationFlags::values, EvaluationFlags::gradients, and EvaluationFlags::hessians are supported."));
```
Can you please break the line at 80 characters?

kronbichler commented on tests/matrix_free/matrix_vector_hessians_faces.cc:
```diff
@@ -0,0 +1,388 @@
+{
+  dealii::Utilities::MPI::MPI_InitFinalize mpi(argc, argv, 1);
+
+  initlog();
```
```suggestion
  initlog();
  deallog << std::setprecision(10);
```
I would add this to be on the safe side for the large numbers appearing here and the relative accuracy of 1e-8 applied by numdiff.

bergbauer replied:
Done!

kronbichler commented on tests/matrix_free/matrix_vector_hessians_faces.cc:
```diff
@@ -0,0 +1,388 @@
+
+  LinearAlgebra::distributed::Vector<double> src, dst, dst2;
+  matrix_free.initialize_dof_vector(src);
+  srand(42);
+  for (auto &v : src)
+    v = static_cast<double>(rand()) / RAND_MAX;
```
```suggestion
  for (double &v : src)
    v = random_value<double>();
```
We use special random numbers rather than `std::rand()` to ensure platform-independence.

bergbauer replied:
I already use the dealii functions here. Should I use `random_value` (which sets the seed to 1) or I can also be more explicit and use `Testing::srand`  and `Testing::rand`?

bergbauer replied:
I applied your suggestion.

kronbichler commented on tests/matrix_free/matrix_vector_hessians_faces.cc:
```diff
@@ -0,0 +1,388 @@
+
+  Triangulation<dim> tria;
+  GridGenerator::hyper_cube(tria);
+  for (int i = 0; i < 3; ++i)
+    tria.refine_global();
```
```suggestion
  tria.refine_global(3);
```
As a side question: Does the code also work if we have hanging nodes, i.e., the first cell or a few cells are refined adaptively? In that case it would make sense to add that here to make the test general.

bergbauer replied:
I tried it out and FEFaceValues and FEFaceEvaluation have different results in this case. Should we address this in this PR?

peterrum replied:
Let's postpone this!

bergbauer replied:
Fine for me. I added a `TODO` to the test.

kronbichler reviewed (commented):
Very nice, thanks for tackling this long-standing issue. I have only a few minor comments and suggestions. Some of them should be done in a different PR in the future, not here.

bergbauer commented:
I addressed all of your comments about `fe_evaluation.h` and `evaluation_kernels.h` @peterrum @kronbichler . We should talk about the remaining issues with hanging nodes and the test.

peterrum commented:
\rebuild

peterrum reviewed (approved):
This looks very good to me. But, let's wait for @kronbichler's last call here!

peterrum commented:
@bergbauer Could you force push again. Jenkins seems to be stuck!
